pkg/controllers/jobset_controller.go

```diff
@@ -913,3 +936,88 @@ func findReplicatedStatus(replicatedJobStatus []jobset.ReplicatedJobStatus, repl
 	}
 	return jobset.ReplicatedJobStatus{}
 }
+
+func (r *JobSetReconciler) deleteJobSet(ctx context.Context, js *jobset.JobSet) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	policy := metav1.DeletePropagationForeground
+	options := []client.DeleteOption{client.PropagationPolicy(policy), client.Preconditions{UID: &js.UID}}
+	log.V(2).Info("Cleaning up JobSet", "jobset", klog.KObj(js))
+
+	return r.Client.Delete(ctx, js, options...)
+}
+
+// checkIfTTLExpired checks whether a given JobSet's TTL has expired.
+func (r *JobSetReconciler) checkIfTTLExpired(ctx context.Context, jobSet *jobset.JobSet) (expiredAt *time.Time, err error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// We don't care about the JobSets that are going to be deleted
+	if jobSet.DeletionTimestamp != nil {
+		return nil, nil
+	}
+
+	now := r.clock.Now()
+	remaining, expiredAt, err := timeLeft(log, jobSet, &now)
+	if err != nil {
+		return nil, err
+	}
+
+	// TTL has expired
+	if *remaining <= 0 {
+		return expiredAt, nil
+	}
+
+	return nil, nil
+}
+
+// timeLeft returns the time left until the JobSet's TTL expires and the time when it will expire.
+func timeLeft(log logr.Logger, js *jobset.JobSet, since *time.Time) (*time.Duration, *time.Time, error) {
+	finishAt, expireAt, err := getJobSetFinishAndExpireTime(js)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	if finishAt.After(*since) {
```

Comment:
We should nil check the pointer before proceeding with the rest of the function which depends on dereferencing it. 

I think each function as a standalone unit of code should be safe from nil dereference errors, and not depend on safety checks elsewhere in the code base, since if we ever change those safety checks it may be easy to miss one place where we depended on them and the controller crashes.

Reply:
added safety checks above